### ISSUE TITLE
Add javac to third_party [phab:D1540]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1958]
_Created by **cushon** at 2017-04-19 06:09:33_: 

Previously kythe was depending on the version of javac that ships with
Bazel. This change makes it easier to upgrade the javac version
independently from Bazel releases. The javac artifact that was added is
the same version that shipped with Bazel 4.5.

This change also updates BUILD files to use the new dependency, and
fixes a few places where javac or javac-based tools weren't using the
-Xbootclasspath/p: flag to prepend javac to the bootclasspath.

**Comment by kythe** [comment]
_**cushon** wrote at 2017-04-19 16:22:51_:

Switch to the version currently provided by Bazel, and also update BUILD files.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-04-19 16:38:52_:

After patching this change locally, I found I'm not able to build, per the following transcript (lightly redacted):

```
$ bazel build //kythe/java/...                                                                                                                                                                                                    
INFO: Found 30 targets...
ERROR: projects/kythe/third_party/javac/BUILD:10:1: Extracting interface //third_party/javac:javac failed: Process terminated by signal 6 [sandboxed].
mmap(): Invalid argument
Unable to open Zip file third_party/javac/javac-9-dev-r3297-4.jar: Invalid argument
Use --strategy=JavaIjar=standalone to disable sandboxing for the failing actions.
INFO: Elapsed time: 1.032s, Critical Path: 0.51s
```

Do we need to update Bazel first?

```
$ bazel version
Build label: 0.4.5-2017-04-12 (037b9b93f)
[…]
```

**Comment by kythe** [comment]
_**cushon** wrote at 2017-04-19 16:42:53_:

>>! In (earlier comment), fromberger wrote:
> After patching this change locally, I found I'm not able to build, per the following transcript (lightly redacted):

arc was complaining about not being able to upload binary artifacts. Does fetching the javac jar again help?

(cd third_party/javac && wget http://repo1.maven.org/maven2/com/google/errorprone/javac/9-dev-r3297-4/javac-9-dev-r3297-4.jar)

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-04-19 16:47:45_:

>>! In (earlier comment), cushon wrote:
>>>! In (earlier comment), fromberger wrote:
>> After patching this change locally, I found I'm not able to build, per the following transcript (lightly redacted):
> 
> arc was complaining about not being able to upload binary artifacts. Does fetching the javac jar again help?
> 
> (cd third_party/javac && wget http://repo1.maven.org/maven2/com/google/errorprone/javac/9-dev-r3297-4/javac-9-dev-r3297-4.jar)

Ah, yes, that helped, thanks. Arcanist apparently created an empty .jar file in that directory, so in addition I had to rename the fetched one (to which wget helpfully added a ".1" to avert a naming collision).

Running all the tests and such now.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-04-19 17:05:58_:

Okay, everything looks good here.

To land this, I'm going to have to manually update the patch to include the JAR; it's possible Arcanist may lose the attribution (if so, I apologize)

**Comment by kythe** [comment]
_**fromberger** accepted this revision at 2017-04-19 17:05:58_

**Comment by kythe** [comment]
_**fromberger** closed via commit 2d585a113cd78e9356594611ec58eac252a79765 at 2017-04-19 17:06:48_
